File: maximal_rectangle.py
class Solution:
    def maximalRectangle(self,matrix):
        if len(matrix)==0:
            return 0
        h=[0 for i in range(len(matrix[0]))]
        maxArea=0
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if matrix[i][j]=='1':
                    h[j]+=1
                else:
                    h[j]=0
            maxArea=max(maxArea,self.largestRectangleArea(h))
        return maxArea
        
    def largestRectangleArea(self,height):
        stack=[]
        height.append(-1)
        length=len(height)
        sumArea=0
        index=0
        while index<length:
            if len(stack)==0 or height[stack[-1]]<=height[index]:
                stack.append(index)
                index+=1
            else:
                top=stack.pop()
                sumArea=max(sumArea,height[top]*(index if len(stack)==0 else index-stack[-1]-1))
        return sumArea


if __name__=='__main__':
    sol=Solution
    matrix=[['0','0','0','0'],['0','0','1','0'],['0','1','1','0'],['1','0','1','1']]
    print sol.maximalRectangle(matrix)

File: Algorithm/CPP/maximal_rectangle.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <algorithm>
using namespace std;


class Solution {
public:
	int maximalRectangle(vector<vector<char>>& matrix) {
		int m = matrix.size();
		if (m < 1) return 0;
		int n = matrix[0].size();

		vector<int> rectangle(n + 1, 0);
		rectangle[n] = -1;

		int res = 0;
		for (int i = 0; i < m; i++)
		{
			for (int j = 0; j < n; j++)
			{
				if (matrix[i][j] == '1') rectangle[j] += 1;
				else rectangle[j] = 0;
			}
			stack<int> Stack;
			int k = 0;


			while (k < n + 1)
			{
				if (Stack.empty() || rectangle[k] >= rectangle[Stack.top()])
				{

					Stack.push(k);
					k++;
				}
				else
				{
					int top = Stack.top();
					Stack.pop();
					if (Stack.empty()) res = max(res, k*rectangle[top]);
					else res = max(res, (k - Stack.top() - 1)*rectangle[top]);
				}
			}
		}
		return res;
	}
};


int main()
{
	Solution sol;
	vector<vector<char>> matrix = {
		{'1','0','1','0','0'},
		{'1','0','1','1','1'},
		{'1','1','1','1','1'},
		{ '1', '0', '0', '1', '0' }
	};
	matrix = { { '1','0', '1', '0' }, { '1','0', '1', '1' }, 
	{ '1', '0', '1', '1' },
	{'1','1','1','1'} };
	//matrix = { { 0, 0 } };
	//matrix = { { '0' } };
	cout << sol.maximalRectangle(matrix) << endl;


	system("pause");
	return 0;
}

